File: ABC BHM/R/ABC_BHM_Documented_Simple_Example.R
# R packages 
#These packages are used for plotting the data
library(ggplot2)
library(ggpubr)
#I use this to get a truncated normal distribution, 
#however I could have just used a beta ditribution
library(truncnorm)
#coda package is used to compute the Gelman & Rubin Statistic
library(coda)
#ABC Sampling Functions
#--------------------------------------------------------------------------------------------------------------------
Sample_Prior           = function(){
#Sample Prior function
#   - ARGUMENTS - takes in no arguments 
#   - RETURNS   - a list of a parameter values samples from the predetermined parameters value
  
Parameters =   rbeta(1, 1, 1) # sample value from beta distribution with flat priors 

#As we develop models with more parameters we will have to modify how we store the parameter values,
# either in a list.
return(Parameters) 
}
Hamiltonion_Sampling   = function(Individual_1_History, Individual_2_History, Sample, Sample_Prev, i){
  #ARGUMENTS: 
  #     - Individual_1_History, This is a list of posterior samples representing the individual subject 1
  #     - Individual_2_History, This is a list of posterior samples representing the individual subject 2
  #     - Sample              , Current population mean parameter 
  #     - Sample_Prev         , Previously accepted population mean parameter
  #     - i                   , represents the iteration of the Markov Chain
   
  #RETURN
  #     - Returns the liklihood ratio for the population parameter

  #Compute the likelihood ratio between the previous sample parameter and the current sample
  #Current likelihood value
  Current_Sample   = sum( dbeta( c(Individual_1_History[i-1], Individual_2_History[i-1]),      Sample * 2, (1 - Sample)      * 2, log = T )  ) 
  #Previous likelihoood value
  Previous_Sample  = sum( dbeta( c(Individual_1_History[i-1], Individual_2_History[i-1]), Sample_Prev * 2, (1 - Sample_Prev) * 2, log = T )  )
  #Compute Likelihood ratio
  Likelihood_Ratio = Current_Sample - Previous_Sample
  return(Likelihood_Ratio)
}
Markov_Chain           = function(Tuning, DV){
  # ARGUMENTS
  #  - Tuning Parameter , Tuning parameter for the ABC sampling algorithm
  #  - DV               , Data that the model is fitting
  #RETURN
  #   -Returns the estimated parameter values for a single Markov Chain
   
  # Markov_Chain funtion runs a single instance of a markov chain of the given model 
  # and is called in the main model function inorder to implement multiple markov Chains
  
  for(i in 1:10000){ #Markov Chain iterates for 10000 times, in the future add Markov chain length as an argument in the function

    print(i)
    
    
    #Estimate the Sample Parameter
    if(i == 1){ #if i is equal to 1 then initialize sample variable from the prior
      
      Sample       = Sample_Prior()
      
    } else { #if i is not equal to 1, then determine the Sample parameter VIA Hamiltonion sampling. 
      
      Sample       =  Sample_Prior()      # Take one sample from Prior distribution
      Sample_Prev  = Sample_History[i-1]  # From the previous value (i-1) Sample_History array 
      
      r = Hamiltonion_Sampling( Individual_1_History, Individual_2_History, Sample, Sample_Prev, i )  # the likelihood ratio between the current and previous value

      # if r (likelihood ratio) is less than randomly sampled value then or equal to the liklihood ratio 
      # then we keep the newly sampled value else we keep the previous Sample Value
      if( log(runif(1, 0,1))  <= r){ Sample } else {Sample = Sample_Prev } 
      
    }
    
    #Sample parameters for Individual participants, keep in mind that for this example we are holding the rate parameter constant 
    Individual_1 = rbeta(1, Sample[1] * 10, (1-Sample[1]) * 10) # Individual_1, Sampled value for individual 1
    Individual_2 = rbeta(1, Sample[1] * 10, (1-Sample[1]) * 10) # Individual_2, Sampled value for indiivdual 2
    
    #-----------------------------------------------------------------------------------------------
    # Once we get this formulation integrated with ACT-R would go here
    
    #--------------------------------Send Parameters to ACT-R here------------------------
    #---------------------------------ACT-R Model Returns behavior------------------------
    #---------Summerize behavior in a meaninful way to compare against participants data
    #-----------------------------------------------------------------------------------------------
    
    #Compute the residual between model predictions and observed data
    Distance_1 = sqrt(  mean(   (DV[1:50]  - Individual_1[1])^2 ) )  # Root Mean Squared Deviation between Individual_1 parameter and Subject 1's Data
    Distance_2 = sqrt(  mean( (DV[51:100]  - Individual_2[1])^2 ) )  # Root Mean Squared Deviation between Individual_2 parameter and Subject 1's Data  
    
    #add together residual from all participants
    Distance = Distance_1 + Distance_2   # Add the distances together
    
    if(i == 1){Distance_Post = 0} #if i == 1 then intialize Distance_Post to 0
    
    #Compute Liklihood ratio to ABC sampling
    alpha    = dnorm(Distance/Tuning, log = T) - dnorm(Distance_Post/Tuning, log = T) # generate the likelihood ratio for individual participants using the Distance variable and the Tuning Parameter
    #Sample random value
    p        = log( runif(1, 0 , 1) ) #Sample random variable to compare against alpha variable
    
    #Save the model results
    if(i == 1){ #if i is 1 then we initialize 4 differnt array that store the results for each markov chain
      Sample_History         = Sample
      Individual_1_History   = Individual_1[1]
      Individual_2_History   = Individual_2[1]
      Distance_Post          = Distance
      
    } else { #if i is not equal to 1, then we store the results based on the p and alpha variable
      
      if(p <= alpha){ # if p is less then alpha then we store the newly sampled parameter values
        Sample_History         = c(Sample_History, Sample )
        Individual_1_History   = c(Individual_1_History, Individual_1[1] ) 
        Individual_2_History   = c(Individual_2_History, Individual_2[1] )
        Distance_Post = Distance  
        
      } else { # if  p is greater and alpha then we store the previously sampled values
        Sample_History         = c(Sample_History      , Sample_History[i-1] )
        Individual_1_History   = c(Individual_1_History, Individual_1_History[i-1] ) 
        Individual_2_History   = c(Individual_2_History, Individual_2_History[i-1] )
        
      }
    }
  
    
  }
  
  #Once the markov chain is finished then we store the results into a data frame 
  Chain           = as.data.frame( cbind(1:length(Sample_History), Sample_History, Individual_1_History, Individual_2_History))
  #rename the columnes of the dataframe
  colnames(Chain) = c("Sample", "Sample_Mean", "Individual_1", "Individual_2")
  #Return the Chain varaible
  return(Chain)
}
ABC_Hierarchical_Model = function(Tuning, DV){
  #This funciton runs the individual markov chains, given the data and tuning parameters 
  #PARAMETERS
  #Tuning parmaeters: Parmaeter controls the likelihood function for the ABC samplers
  #DV: the Data for the Individual participants
  #Note: this could be paralized in the future
  #RETURNS
  # The results from the ABC BHM Model
  Chain_1 = Markov_Chain(Tuning, DV)  # Run chain 1
  Chain_2 = Markov_Chain(Tuning, DV)  # Run chain 2
  Chain_3 = Markov_Chain(Tuning, DV)  # Run chain 3
  
  #label the Chains
  Chain_1$Chain = 1 
  Chain_2$Chain = 2
  Chain_3$Chain = 3
  
  #Rbind the data together
  Data = rbind(Chain_1, Chain_2, Chain_3) 
  #Return the data together.s
  return(Data)  
}
#--------------------------------------------------------------------------------------------------------------------
#Generate the data
Sub  = c( rep(1, 50) , rep(2, 50))
DV   = c( rep(.3, 50), rep(.7, 50) )
DV   = c( rtruncnorm(50, .3, .05, a= 0,b=1), rtruncnorm(50, .7, .05, a= 0, b = 1) )
Obs  = as.data.frame( cbind(Sub, DV) )
#--------------------------------------------------------------------------------------------------------------------
# Run the Various Model Simulation manipulating the the Turing parametetr
Tuning_05 = ABC_Hierarchical_Model(.05, DV)
Tuning_10 = ABC_Hierarchical_Model(.10, DV)
Tuning_15 = ABC_Hierarchical_Model(.15, DV)
Tuning_20 = ABC_Hierarchical_Model(.20, DV)
#--------------------------------------------------------------------------------------------------------------------
#Observse the results
#Here is the Simulated Data
{
ggplot(Obs, aes(DV, group = Sub) ) +
  geom_density(aes(fill = as.factor(Sub) ) ) + 
  xlim(0,1) +
  xlab("Ability") + ylab("Density") +
  labs(fill = "Simulated Data")  +
  ggtitle("Generated Data" ) + 
  theme(
    title      = element_text(size = 16),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12)
  )

}
#--------------------------------------------------------------------------------------------------------------------
#Here we observe the Markov Chain of the Various simulations
{
title_size      = 14
axis_title_size = 14
axis_text_size  = 12

Sample_Mean_Tau_05 = ggplot(Tuning_05[Tuning_05$Sample > 5000, ], aes(Sample, Sample_Mean, group = Chain) ) +
  geom_line( aes(col = as.factor(Chain) )) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Tau .05: Population Mean") +
  labs(col = "Chain") +
  ylim(0,1) +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )

Sample_Mean_Tau_10 = ggplot(Tuning_10[Tuning_10$Sample > 5000, ], aes(Sample, Sample_Mean, group = Chain) ) +
  geom_line( aes(col = as.factor(Chain) )) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Tau .1: Population Mean") +
  labs(col = "Chain") +
  ylim(0,1) +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )

Sample_Mean_Tau_15 = ggplot(Tuning_15[Tuning_15$Sample > 5000, ], aes(Sample, Sample_Mean, group = Chain) ) +
  geom_line( aes(col = as.factor(Chain) )) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Tau .15: Population Mean") +
  labs(col = "Chain") +
  ylim(0,1) +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )


Sample_Mean_Tau_20 = ggplot(Tuning_20[Tuning_20$Sample > 5000, ], aes(Sample, Sample_Mean, group = Chain) ) +
  geom_line( aes(col = as.factor(Chain) )) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Tau .20: Population Mean") +
  labs(col = "Chain") +
  ylim(0,1) +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )


ggarrange(Sample_Mean_Tau_05, 
          Sample_Mean_Tau_10, 
          Sample_Mean_Tau_15, 
          Sample_Mean_Tau_20,
          ncol = 2, nrow = 2)
}
#--------------------------------------------------------------------------------------------------------------------------------------------
#Here we look at the estiamted posterior distribution
{
title_size      = 14
axis_title_size = 14
axis_text_size  = 12

Posterior_Tau_05 = ggplot(Tuning_05[Tuning_05$Sample > 5000, ], aes( Sample_Mean) ) +
  #geom_density( aes( fill = "Posterior_Mean" ) ) +
  geom_density( aes(Individual_1, fill = "Posterior Subj 1" ), alpha = .25 ) +
  geom_density( aes(Individual_2, fill = "Posterior Subj 2" ), alpha = .25 ) +
  geom_density(data = Obs, aes(DV, group = as.factor(Sub), fill = "Data"), alpha = .25) +
  labs(fill = "Parameter") +
  xlim(0,1) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Posterior Samples") +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )


Posterior_Tau_10 = ggplot(Tuning_10[Tuning_10$Sample > 5000, ], aes( Sample_Mean) ) +
  #geom_density( aes( fill = "Posterior_Mean" ) ) +
  geom_density( aes(Individual_1, fill = "Posterior Subj 1" ), alpha = .25 ) +
  geom_density( aes(Individual_2, fill = "Posterior Subj 2" ), alpha = .25 ) +
  geom_density(data = Obs, aes(DV, group = as.factor(Sub), fill = "Data"), alpha = .25) +
  labs(fill = "Parameter") +
  xlim(0,1) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Posterior Samples") +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )

Posterior_Tau_15 = ggplot(Tuning_15[Tuning_15$Sample > 5000, ], aes( Sample_Mean) ) +
  #geom_density( aes( fill = "Posterior_Mean" ) ) +
  geom_density( aes(Individual_1, fill = "Posterior Subj 1" ), alpha = .25 ) +
  geom_density( aes(Individual_2, fill = "Posterior Subj 2" ), alpha = .25 ) +
  geom_density(data = Obs, aes(DV, group = as.factor(Sub), fill = "Data"), alpha = .25) +
  labs(fill = "Parameter") +
  xlim(0,1) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Posterior Samples") +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )

Posterior_Tau_20 = ggplot(Tuning_20[Tuning_20$Sample > 5000, ], aes( Sample_Mean) ) +
  #geom_density( aes( fill = "Posterior_Mean" ) ) +
  geom_density( aes(Individual_1, fill = "Posterior Subj 1" ), alpha = .25 ) +
  geom_density( aes(Individual_2, fill = "Posterior Subj 2" ), alpha = .25 ) +
  geom_density(data = Obs, aes(DV, group = as.factor(Sub), fill = "Data"), alpha = .25) +
  labs(fill = "Parameter") +
  xlim(0,1) +
  xlab("Sample") + ylab("Value")+
  ggtitle("Posterior Samples") +
  theme(
    title      = element_text(size = title_size),
    axis.title = element_text(size = axis_title_size),
    axis.text  = element_text(size = axis_text_size)
  )

ggarrange(Posterior_Tau_05, 
          Posterior_Tau_10, 
          Posterior_Tau_15, 
          Posterior_Tau_20,
          ncol = 2, nrow = 2)
}
#-----------------------------------------------------------------------------------------------------------------------------
{
#Here we assess the convergence of the Markov Chain

Tau_05 = mcmc.list(
  as.mcmc(Tuning_05$Sample_Mean[Tuning_05$Chain == 1]),
  as.mcmc(Tuning_05$Sample_Mean[Tuning_05$Chain == 2]),
  as.mcmc(Tuning_05$Sample_Mean[Tuning_05$Chain == 3])
)

Tau_10 = mcmc.list(
  as.mcmc(Tuning_10$Sample_Mean[Tuning_15$Chain == 1]),
  as.mcmc(Tuning_10$Sample_Mean[Tuning_15$Chain == 2]),
  as.mcmc(Tuning_10$Sample_Mean[Tuning_15$Chain == 3])
)


Tau_15 = mcmc.list(
  as.mcmc(Tuning_15$Sample_Mean[Tuning_15$Chain == 1]),
  as.mcmc(Tuning_15$Sample_Mean[Tuning_15$Chain == 2]),
  as.mcmc(Tuning_15$Sample_Mean[Tuning_15$Chain == 3])
)

Tau_20 = mcmc.list(
  as.mcmc(Tuning_20$Sample_Mean[Tuning_15$Chain == 1]),
  as.mcmc(Tuning_20$Sample_Mean[Tuning_15$Chain == 2]),
  as.mcmc(Tuning_20$Sample_Mean[Tuning_15$Chain == 3])
)


GD_05 = gelman.diag(Tau_05) 
GD_10 = gelman.diag(Tau_10)
GD_15 = gelman.diag(Tau_15)
GD_20 = gelman.diag(Tau_20)


Gelman_Diag = data.frame(
  Tau = c(.05, .1, .15, .2),
  Point.est = c(GD_05$psrf[1], GD_10$psrf[1],GD_15$psrf[1],GD_20$psrf[1]),
  Upper_CI = c(GD_05$psrf[2], GD_10$psrf[2],GD_15$psrf[2],GD_20$psrf[2])
)

ggplot(Gelman_Diag, aes(Tau, Point.est) ) +
  geom_line(aes(col = "Point.est") ) +
  geom_point(aes(col = "Point.est") )  +
  geom_line( aes(Tau, Upper_CI, col = "Upper CI") ) +
  geom_point(aes(Tau, Upper_CI, col = "Upper CI") ) +
  labs(col = "Gelman & Rubin Statistic") +
  xlab("Tau Value") + ylab("R Hat (Lower is better)") +
  ggtitle("Markov Chain Diagonstic") +
  theme(
    title = element_text(size = 16),
    axis.title = element_text(size = 14),
    axis.text  = element_text(size = 12)
  )
}

File: ABC BHM/py/ABC_BHM_Documented_Simple_Example.py
import math
import numpy as np
import pandas as pd
from scipy.stats import beta
from scipy.stats import norm
from scipy.stats import uniform
from scipy.stats import truncnorm

def sample_prior():
    """Returns a list of parameter value samples from a beta distribution with flat priors"""
    parameters = beta.rvs(1,1,size=1)
    return parameters

def hammy_sammy(indiv1Hist, indiv2Hist, sample, samplePrev, i):
    return hamiltonian_sampling(indiv1Hist, indiv2Hist, sample, samplePrev, i)

def hamiltonian_sampling(indiv1Hist, indiv2Hist, sample, samplePrev, i):
    """
    Computes the likelihood ratio for the population parameter.

    Parameters
    ----------
    indiv1Hist : A list of posterior samples representing an individual.
    indiv2Hist : A list of posterior samples representing a second individual.
    sample     : The current population mean parameter.
    samplePrev : The previously accepted population mean parameter.
    i          : Index of the iteration of the Markov Chain.

    Returns
    -------
    llhRatio   : The likelihood ratio for the population parameter.

    """
    
    # compute the likelihood ratio between the previous sample of the parameter and the current sample 
    # kappa = 2 (fixed; only estimating mu)
    currSample = sum(beta.logpdf([indiv1Hist[i-1],indiv2Hist[i-1]], sample*2, (1-sample)*2)) # current likelihood value
    prevSample = sum(beta.logpdf([indiv1Hist[i-1],indiv2Hist[i-1]], samplePrev*2, (1-samplePrev)*2)) # previous likelihood value
    llhRatio   = currSample - prevSample # likelihood ratio
    return llhRatio

def markov_chain(tuning, DV):
    """
    Runs a single instance of a Markov Chain for a given model.

    Parameters
    ----------
    tuning : Tuning parameter for the ABC sampling algorithm.
    DV     : Dependent variable; the data that the model is being fit to.

    Returns
    -------
    chain : The estimated model parameter values from a single Markov Chain.

    """
    
    sampleHist = []
    indiv1Hist = []
    indiv2Hist = []
    for i in range (0,10000):
        print("Run number "+str(i))
        
        # estimate the sample parameter
        if i==0: # initialize sample parameter from the prior
            sample = sample_prior()
        else: # determine sample parameter via Hamiltonian sampling
            sample = sample_prior()
            samplePrev = sampleHist[i-1]
            
            # get the likelihood ratio between the current and previous samples
            r = hammy_sammy(indiv1Hist,indiv2Hist,sample,samplePrev,i)
            
            # if the likelihood ratio is less than or equal to some randomly sampled value,
            # keep the newly sampled parameter value; otherwise, keep the previously sampled parameter value
            if math.log(uniform.rvs(0,1,1)) <= r: # check
                pass
            else:
                sample = [samplePrev]
                
        # sample parameters for individual participants
        # in this example the rate parameter is held constant
        indiv1 = beta.rvs(sample[0]*10, (1-sample[0])*10, size=1)
        indiv2 = beta.rvs(sample[0]*10, (1-sample[0])*10, size=1)
            
        ###
        # integrate ACT-R here
        # (probably) don't have to integrate behavioral measures! 
        # just calculate the residual for each measure, then when alpha is checked against p, it has to pass checks
        # for each of the behavioral measures to be kept
        ###
            
        #compute residual between model predictions and observed data
        dist1 = np.sqrt(np.mean((DV[0:50] - indiv1[0])**2))
        dist2 = np.sqrt(np.mean((DV[51:100] - indiv2[0])**2))
            
        #add residuals from all participants
        dist = dist1 + dist2
            
        # initialize distPost to 0
        if i==0:
            distPost = 0.01
            
        # compute likelihood ratio for ABC sampling
        #alpha = norm.logpdf(dist/tuning) - norm.logpdf(distPost/tuning)
        alpha = math.log(norm.pdf(dist/tuning)) - math.log(norm.pdf(distPost/tuning))
        # sample a random value to test against
        p = math.log(uniform.rvs(0,1,1))
            
        if i==0:
            sampleHist.append(sample[0])
            indiv1Hist.append(indiv1[0])
            indiv2Hist.append(indiv2[0])
            distPost = dist
        else:
            if p <= alpha: # if p is less than or equal to alpha, store newly sampled values
                sampleHist.append(sample[0])
                indiv1Hist.append(indiv1[0])
                indiv2Hist.append(indiv2[0])
                distPost = dist
            else: # otherwise store the previously sampled values again
                sampleHist.append(sampleHist[i-1])
                indiv1Hist.append(indiv1Hist[i-1])
                indiv2Hist.append(indiv2Hist[i-1])
                    
    # store chain as dict
    chain = {"Sample":list(range(0,len(sampleHist))),
             "SampleMean":sampleHist,
             "Individual_1":indiv1Hist,
             "Individual_2":indiv2Hist}
    return chain


def abc_hierarchical_model(tuning, dv):
    """
    Runs a set of Markov Chains for a given model.

    Parameters
    ----------
    tuning : Tuning parameter for the ABC sampling algorithm.
    DV     : Dependent variable; the data that the model is being fit to.

    Returns
    -------
    data : The results of the ABC hierarchical sampling model.

    """    
    chain1 = pd.DataFrame(markov_chain(tuning,dv))
    chain2 = pd.DataFrame(markov_chain(tuning,dv))
    chain3 = pd.DataFrame(markov_chain(tuning,dv))
    
    chain1['Chain'] = 1
    chain2['Chain'] = 2
    chain3['Chain'] = 3
    
    data = pd.concat([chain1,chain2,chain3],ignore_index=True)
    return data



# generate data
part = [1]*50 + [2]*50
dv = [.3]*50 + [.7]*50
dv = list(truncnorm.rvs(a=0, b=1, loc=0.3, scale=0.05, size=50)) + list(truncnorm.rvs(a=0, b=1, loc=0.7, scale=0.05, size=50))

partData = {"partID":part,
            "dvSteady":[.3]*50 + [.7]*50,
            "dv":dv
    }

tuning05 = abc_hierarchical_model(.05,dv)
tuning10 = abc_hierarchical_model(.1,dv)
tuning15 = abc_hierarchical_model(.15,dv)   
tuning20 = abc_hierarchical_model(.2,dv)

tuning05.to_csv('/Users/pjr5/Desktop/BCB-Modeling-main/ABC BHM R Code/pyData/tuning05.csv')    
tuning10.to_csv('/Users/pjr5/Desktop/BCB-Modeling-main/ABC BHM R Code/pyData/tuning10.csv')    
tuning15.to_csv('/Users/pjr5/Desktop/BCB-Modeling-main/ABC BHM R Code/pyData/tuning15.csv')    
tuning20.to_csv('/Users/pjr5/Desktop/BCB-Modeling-main/ABC BHM R Code/pyData/tuning20.csv')    

partData = pd.DataFrame(partData)
partData.to_csv('/Users/pjr5/Desktop/BCB-Modeling-main/ABC BHM R Code/pyData/partData.csv')
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    